logs_intake/socket/SendLogEntriesOverSocket.py

```diff
@@ -0,0 +1,104 @@
+
+
+def lambda_handler(event):
+    # Check prerequisites
+    if ddApiKey == "<your_api_key>" or ddApiKey == "":
+        raise Exception(
+            "You must configure your API key before starting this lambda function (see #Parameters section)"
+        )
+
+    # Attach Datadog's Socket
+    s = connect_to_datadog(host, ssl_port)
+
+    try:
+        s = safe_submit_log(s, event)
+    except Exception as e:
+        print('Unexpected exception: {} for event {}'.format(str(e), event))
+    finally:
+        s.close()
+
+def connect_to_datadog(host, port):
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    s = ssl.wrap_socket(s)
+    s.connect((host, port))
+    return s
+
+def safe_submit_log(s, log):
+    try:
+        send_entry(s, log)
+    except Exception as e:
+        # retry once
+        s = connect_to_datadog(host, ssl_port)
+        send_entry(s, log)
+    return s
+
+# Utility functions
+def send_entry(s, log_entry):
+    # The log_entry can only be a string or a dict
+    if isinstance(log_entry, str):
+        log_entry = {"message": log_entry}
+    elif not isinstance(log_entry, dict):
+        raise Exception(
+            "Cannot send the entry as it must be either a string or a dict. Provided entry: " + str(log_entry)
+        )
+    # Merge with metadata
+    log_entry = merge_dicts(log_entry, metadata)
+    print("log_entry has value {}".format(log_entry))
+
+    # Send to Datadog
+    str_entry = json.dumps(log_entry)
+    print(str_entry)
+    prefix = "%s " % ddApiKey
+    return s.send((prefix + str_entry + "\n").encode("UTF-8"))
+
+def merge_dicts(a, b, path=None):
+    if path is None:
+        path = []
+    for key in b:
+        if key in a:
+            if isinstance(a[key], dict) and isinstance(b[key], dict):
+                merge_dicts(a[key], b[key], path + [str(key)])
+            elif a[key] == b[key]:
+                pass  # same leaf value
+            else:
+                raise Exception(
+                    'Conflict while merging metadatas and the log entry at %s' % '.'.join(path + [str(key)])
+                )
+        else:
+            a[key] = b[key]
+    return a
+
+if __name__ == "__main__":
+    # execute only if run as a script
+    metadata[DD_CUSTOM_TAGS] = "x.error,level:error,x.api-identity-v1,x.api:identity-v1,x.env-prd,x.env:prd,x.proj-apigee,x.proj:apigee"
+    lambda_handler({"apiProduct":"product-v1-prod","appName":"morrisons-xxwmm-basketgateway_service-prod","clientIp":"52.49.111.54","environment":"prd","clientRequestResource":"/items/314","clientRequestSize":0,"clientRequestUri":"/product/v1/items/314","clientRequestQuerystring":"fields=itemDescription%2CcommercialHierarchy&apikey=m6jR90jCJtASGUArO4vik5BpsFTVIEj6","clientRequestVerb":"GET","clientRequestTime":1524836354312,"responseCode":404,"backendResponseCode":404,"timeBackend":5,"timeTotal":10,"errorMessage":"Item not found","errorCode":"404.19.001","userAgent":"AmazonSQSBufferedAsyncClient/1.11.123"})
```

Comment:
it appears as though there are some potentially sensitive strings in these example arguments? like api key fields? we should probably scrub this... in fact, this is a public repo, so you may have already leaked something even by pushing this branch...

this and the last comment apply to the next few lines as well.

Reply:
API keys and customer name removed.  I don't know how to check if there is any history that also needs removing.